src/Processors/QueryPlan/Optimizations/removeRedundantDistinct.cpp

```diff
@@ -0,0 +1,148 @@
+#include <memory>
+#include <Interpreters/ActionsDAG.h>
+#include <Processors/QueryPlan/DistinctStep.h>
+#include <Processors/QueryPlan/ExpressionStep.h>
+#include <Processors/QueryPlan/FilterStep.h>
+#include <Processors/QueryPlan/IntersectOrExceptStep.h>
+#include <Processors/QueryPlan/JoinStep.h>
+#include <Processors/QueryPlan/Optimizations/Optimizations.h>
+#include <Processors/QueryPlan/UnionStep.h>
+#include <Common/logger_useful.h>
+#include <Common/typeid_cast.h>
+
+namespace DB::QueryPlanOptimizations
+{
+
+namespace
+{
+    constexpr bool debug_logging_enabled = false;
+    void logActionsDAG(const String & prefix, const ActionsDAGPtr & actions)
+    {
+        if constexpr (debug_logging_enabled)
+        {
+            LOG_DEBUG(&Poco::Logger::get("redundantDistinct"), "{}: {}", prefix, actions->dumpDAG());
+        }
+    }
+
+    std::set<std::string_view> getDistinctColumns(const DistinctStep * distinct)
```

Comment:
Maybe it is okay here but passing pointers around is usually smelly (as they can theoretically be nullptr). Better use references.